pkg/admission/webhook/generic_webhook.go

```diff
@@ -39,13 +39,52 @@ import (
 
 const byWorkspaceIndex = "webhookDispatcher-byWorkspace"
 
+type ClusterAwareSource interface {
+	Webhooks(cluster logicalcluster.Name) []webhook.WebhookAccessor
+	HasSynced() bool
+}
+
+type clusterAwareSource struct {
+	factory   func(cluster logicalcluster.Name) generic.Source
+	hasSynced func() bool
+
+	lock    sync.RWMutex
+	sources map[logicalcluster.Name]generic.Source
+}
+
+func (c *clusterAwareSource) Webhooks(cluster logicalcluster.Name) []webhook.WebhookAccessor {
+	var source generic.Source
+	var found bool
+	c.lock.RLock()
+	source, found = c.sources[cluster]
+	c.lock.RUnlock()
+	if found {
+		return source.Webhooks()
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	source, found = c.sources[cluster]
+	if found {
+		return source.Webhooks()
+	}
+
+	source = c.factory(cluster)
+	c.sources[cluster] = source
```

Comment:
how is this map ever being cleaned up?

Reply:
It's not - seems like a good follow-up. We need the rebase to 1.25 to get resource event handler function cleanup on the (shared, cross-cluster) list.